builtin/method_str.py

```diff
@@ -26,100 +26,184 @@
+            else:
+                raise error.TypeErr(pattern_val,
+                                    'expected pattern to be Eggex or Str',
+                                    rd.LeftParenToken())
         rd.Done()
 
-        left = 0
-        right = len(string)
-        if self.trim & TRIM_LEFT:
-            left = self._Left(string)
-        if self.trim & TRIM_RIGHT:
-            right = self._Right(string)
+        matched = False
+        try:
+            if pattern_str is not None:
+                if self.anchor & START:
+                    matched, _, _ = string_ops.StartsWithStrByteRange(
+                        string, pattern_str)
+                else:
+                    matched, _, _ = string_ops.EndsWithStrByteRange(
+                        string, pattern_str)
+            else:
+                assert pattern_eggex is not None
+                if self.anchor & START:
+                    matched, _, _ = _StartsWithEggexByteRange(
+                        string, pattern_eggex)
+                else:
+                    matched, _, _ = _EndsWithEggexByteRange(
+                        string, pattern_eggex)
+        except error.Strict as e:
+            raise error.Expr(e.msg, e.location)
+
+        return value.Bool(matched)
 
-        res = string[left:right]
+
+class Trim(vm._Callable):
+    """ Implements `trimStart()`, `trimEnd()`, and `trim()` """
+
+    def __init__(self, anchor):
+        # type: (int) -> None
+        self.anchor = anchor
+
+    def Call(self, rd):
+        # type: (typed_args.Reader) -> value_t
+        """
+        string => trimStart()               # => Str
+        string => trimEnd()                 # => Str
+        string => trim()                    # => Str
+        string => trimStart(pattern_str)    # => Str
+        string => trimEnd(pattern_str)      # => Str
+        string => trim(pattern_str)         # => Str
+        string => trimStart(pattern_eggex)  # => Str
+        string => trimEnd(pattern_eggex)    # => Str
+        string => trim(pattern_eggex)       # => Str
+        """
+
+        string = rd.PosStr()
+        pattern_val = rd.OptionalValue()
+        pattern_str = None  # type: str
+        pattern_eggex = None  # type: value.Eggex
+        if pattern_val:
+            with tagswitch(pattern_val) as case:
+                if case(value_e.Eggex):
+                    pattern_eggex = cast(value.Eggex, pattern_val)
+                elif case(value_e.Str):
+                    pattern_str = cast(value.Str, pattern_val).s
+                else:
+                    raise error.TypeErr(pattern_val,
+                                        'expected pattern to be Eggex or Str',
+                                        rd.LeftParenToken())
+        rd.Done()
+
+        start = 0
+        end = len(string)
+        try:
+            if pattern_str is not None:
+                if self.anchor & START:
+                    _, _, start = string_ops.StartsWithStrByteRange(
```

Comment:
Same comment here -- I think we should skip decoding when trimming a constant prefix or suffix

It can just test for byte equality, and trim if so

Reply:
Ack. Done.